File: uikit/src/main/java/io/agora/chat/uikit/conversation/viewholder/EaseConversationViewHolder.java
package io.agora.chat.uikit.conversation.viewholder;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.view.View;

import androidx.annotation.NonNull;

import com.bumptech.glide.Glide;

import java.util.Date;

import io.agora.chat.ChatClient;
import io.agora.chat.ChatMessage;
import io.agora.chat.ChatRoom;
import io.agora.chat.Conversation;
import io.agora.chat.Group;
import io.agora.chat.uikit.EaseUIKit;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.conversation.model.EaseConversationInfo;
import io.agora.chat.uikit.manager.EaseAtMessageHelper;
import io.agora.chat.uikit.manager.EasePreferenceManager;
import io.agora.chat.uikit.models.EaseConvSet;
import io.agora.chat.uikit.models.EaseUser;
import io.agora.chat.uikit.provider.EaseConversationInfoProvider;
import io.agora.chat.uikit.provider.EaseUserProfileProvider;
import io.agora.chat.uikit.utils.EaseDateUtils;
import io.agora.chat.uikit.utils.EaseSmileUtils;
import io.agora.chat.uikit.utils.EaseUtils;

public class EaseConversationViewHolder extends EaseBaseConversationViewHolder{
    public EaseConversationViewHolder(@NonNull View itemView) {
        super(itemView);
    }

    @Override
    public void setData(EaseConversationInfo bean, int position) {
        super.setData(bean, position);
        Conversation item = (Conversation) bean.getInfo();
        Context context = itemView.getContext();
        String username = item.conversationId();
        mentioned.setVisibility(View.GONE);
        int defaultAvatar = 0;
        String showName = null;
        if(item.getType() == Conversation.ConversationType.GroupChat) {
            if(EaseAtMessageHelper.get().hasAtMeMsg(username)) {
                mentioned.setText(R.string.ease_chat_were_mentioned);
                mentioned.setVisibility(View.VISIBLE);
            }
            defaultAvatar = R.drawable.ease_default_group_avatar;
            Group group = ChatClient.getInstance().groupManager().getGroup(username);
            showName = group != null ? group.getGroupName() : username;
        }else if(item.getType() == Conversation.ConversationType.ChatRoom) {
            defaultAvatar = R.drawable.ease_default_group_avatar;
            ChatRoom chatRoom = ChatClient.getInstance().chatroomManager().getChatRoom(username);
            showName = chatRoom != null && !TextUtils.isEmpty(chatRoom.getName()) ? chatRoom.getName() : username;
        }else {
            defaultAvatar = R.drawable.ease_default_avatar;
            showName = username;
        }
        avatar.setImageResource(defaultAvatar);
        this.name.setText(showName);
        EaseConversationInfoProvider infoProvider = EaseUIKit.getInstance().getConversationInfoProvider();
        if(infoProvider != null) {
            EaseConvSet convSet = infoProvider.getConversationInfo(username);
            if(convSet != null) {
                String name = convSet.getName();
                if(!TextUtils.isEmpty(name)) {
                    this.name.setText(name);
                }
                Drawable icon = convSet.getIcon();
                if(icon != null) {
                    Glide.with(mContext).load(icon).error(defaultAvatar).into(avatar);
                }
                Drawable background = convSet.getBackground();
                if(background != null) {
                    itemView.setBackground(background);
                }
            }
        }
        // add judgement for conversation type
        if(item.getType() == Conversation.ConversationType.Chat) {
            EaseUserProfileProvider userProvider = EaseUIKit.getInstance().getUserProvider();
            if(userProvider != null) {
                EaseUser user = userProvider.getUser(username);
                if(user != null) {
                    if(!TextUtils.isEmpty(user.getNickname())) {
                        this.name.setText(user.getNickname());
                    }
                    if(!TextUtils.isEmpty(user.getAvatar())) {
                        Drawable drawable = avatar.getDrawable();
                        Glide.with(mContext)
                                .load(user.getAvatar())
                                .error(drawable)
                                .into(avatar);
                    }
                }
            }
        }
        // TODO: 2021/10/25 add configures
        if(true) {
            showUnreadNum(item.getUnreadMsgCount());
        }

        if(item.getAllMsgCount() != 0) {
            ChatMessage lastMessage = item.getLastMessage();
            message.setText(EaseSmileUtils.getSmiledText(context, EaseUtils.getMessageDigest(lastMessage, context)));
            time.setText(EaseDateUtils.getTimestampString(context, new Date(lastMessage.getMsgTime())));
            if (lastMessage.direct() == ChatMessage.Direct.SEND && lastMessage.status() == ChatMessage.Status.FAIL) {
                mMsgState.setVisibility(View.VISIBLE);
            } else {
                mMsgState.setVisibility(View.GONE);
            }
        }

        if(mentioned.getVisibility() != View.VISIBLE) {
            String unSendMsg = EasePreferenceManager.getInstance().getUnSendMsgInfo(username);
            if(!TextUtils.isEmpty(unSendMsg)) {
                mentioned.setText(R.string.ease_chat_were_not_send_msg);
                message.setText(unSendMsg);
                mentioned.setVisibility(View.VISIBLE);
            }
        }
    }
}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/feature/conversation/viewholders/EaseConversationViewHolder.kt
package io.agora.uikit.feature.conversation.viewholders

import android.graphics.drawable.Drawable
import android.view.View
import androidx.viewbinding.ViewBinding
import coil.load
import io.agora.uikit.EaseIM
import io.agora.uikit.R
import io.agora.uikit.base.EaseBaseRecyclerViewAdapter
import io.agora.uikit.common.ChatMessageDirection
import io.agora.uikit.common.ChatMessageStatus
import io.agora.uikit.common.ChatType
import io.agora.uikit.common.EaseConstant
import io.agora.uikit.common.extensions.getChatroomName
import io.agora.uikit.common.extensions.getEmojiText
import io.agora.uikit.common.extensions.getGroupNameFromId
import io.agora.uikit.common.extensions.getMessageDigest
import io.agora.uikit.common.extensions.getTextHeight
import io.agora.uikit.common.extensions.getDateFormat
import io.agora.uikit.common.extensions.getUserInfo
import io.agora.uikit.common.extensions.isAlertMessage
import io.agora.uikit.common.helper.EaseAtMessageHelper
import io.agora.uikit.common.helper.EasePreferenceManager
import io.agora.uikit.configs.setAvatarStyle
import io.agora.uikit.databinding.EaseItemConversationListBinding
import io.agora.uikit.feature.conversation.config.EaseConvItemConfig
import io.agora.uikit.feature.conversation.config.bindView
import io.agora.uikit.feature.conversation.config.showUnreadCount
import io.agora.uikit.model.EaseConversation
import io.agora.uikit.model.isChatRoom
import io.agora.uikit.model.isGroupChat
import io.agora.uikit.provider.getSyncProfile
import io.agora.uikit.provider.getSyncUser

class EaseConversationViewHolder(
    private val viewBinding: EaseItemConversationListBinding,
    var config: EaseConvItemConfig? = EaseConvItemConfig()
): EaseBaseRecyclerViewAdapter.ViewHolder<EaseConversation>(binding = viewBinding) {
    private var bgDrawable: Drawable? = null

    init {
        config?.bindView(viewBinding)
    }

    override fun initView(viewBinding: ViewBinding?) {
        super.initView(viewBinding)
        viewBinding?.let {
            if (it is EaseItemConversationListBinding) {
                EaseIM.getConfig()?.avatarConfig?.setAvatarStyle(it.avatar)
                bgDrawable = it.root.background
            }
        }
    }

    override fun setData(item: EaseConversation?, position: Int) {
        item?.let {
            with(viewBinding) {
                // Rest mentioned view's status.
                mentioned.visibility = View.GONE
                ivTopLabel.visibility = if (item.isPinned) View.VISIBLE else View.GONE
                msgMute.visibility = if (item.isSilent()) View.VISIBLE else View.GONE
                item.setOnSelectedListener{ isSelected->
                    if (isSelected) {
                        itemView.setBackgroundResource(R.drawable.ease_conv_item_selected)
                    } else {
                        if (item.isPinned) {
                            itemView.setBackgroundResource(R.drawable.ease_conv_item_pinned)
                        } else {
                            itemView.background = bgDrawable
                        }
                    }
                }
                // If conversation pinned, set background.
                if (item.isPinned) {
                    itemView.setBackgroundResource(R.drawable.ease_conv_item_pinned)
                } else {
                    itemView.background = bgDrawable
                }
                // Set conversation avatar and name.
                var placeholderRes = R.drawable.ease_default_avatar
                if (item.isGroupChat()) {
                    placeholderRes = R.drawable.ease_default_group_avatar
                    avatar.setImageResource(placeholderRes)
                    name.text = item.conversationId.getGroupNameFromId()
                    // Set custom data provided by user
                    EaseIM.getGroupProfileProvider()?.getSyncProfile(item.conversationId)?.let { profile ->
                        avatar.load(profile.avatar) {
                            placeholder(placeholderRes)
                            error(placeholderRes)
                        }
                        if (profile.name.isNullOrEmpty()) {
                            name.text = profile.name
                        }
                    }
                } else if (item.isChatRoom()) {
                    placeholderRes = R.drawable.ease_default_chatroom_avatar
                    avatar.setImageResource(placeholderRes)
                    name.text = item.conversationId.getChatroomName()
                } else {
                    name.text = item.conversationId
                    avatar.setImageResource(placeholderRes)
                    EaseIM.getUserProvider()?.getSyncUser(item.conversationId)?.let { user ->
                        avatar.load(user.avatar) {
                            placeholder(placeholderRes)
                            error(placeholderRes)
                        }
                        name.text = user.getRemarkOrName()
                    }
                }




                // Set @ message in group chat
                if (item.isGroupChat()) {
                    if (EaseAtMessageHelper.get().hasAtMeMsg(item.conversationId)) {
                        mentioned.setText(R.string.ease_chat_were_mentioned)
                        item.lastMessage?.run {
                            getStringAttribute(EaseConstant.MESSAGE_ATTR_AT_MSG, null)?.let { tag ->
                                if (tag == EaseConstant.MESSAGE_ATTR_VALUE_AT_MSG_ALL) {
                                    mentioned.setText(R.string.ease_chat_were_mentioned_all)
                                }
                            }
                        }
                        mentioned.visibility = View.VISIBLE
                    }
                }

                config?.showUnreadCount(item.unreadMsgCount, item.isSilent(), this)

                // Set latest message and failed status
                item.lastMessage?.let {
                    if (it.chatType != ChatType.Chat) {
                        if (it.isAlertMessage()) {
                            message.text = it.getMessageDigest(itemView.context)
                                .getEmojiText(itemView.context, message.getTextHeight())
                        } else {
                            val name = it.getUserInfo()?.getRemarkOrName() ?: it.from
                            message.text = name.plus(": ").plus(it.getMessageDigest(itemView.context)
                                .getEmojiText(itemView.context, message.getTextHeight()))
                        }
                    } else {
                        message.text = it.getMessageDigest(itemView.context)
                            .getEmojiText(itemView.context, message.getTextHeight())
                    }
                    time.text = it.getDateFormat()
                    if (it.direct() == ChatMessageDirection.SEND && it.status() == ChatMessageStatus.FAIL) {
                        msgState.visibility = View.VISIBLE
                    } else {
                        msgState.visibility = View.GONE
                    }
                }?:kotlin.run {
                    message.text = ""
                }

                // Set un-send message info
                if (mentioned.visibility != View.VISIBLE) {
                    EasePreferenceManager.getInstance().getUnSendMsgInfo(item.conversationId)?.let { info ->
                        if (info.isNotEmpty()) {
                            mentioned.setText(R.string.ease_chat_were_not_send_msg)
                            message.text = info
                            mentioned.visibility = View.VISIBLE
                        }
                    }
                }
            }
        }
    }
}